src/js/historical.cpp

```diff
@@ -23,8 +23,34 @@ namespace ccf::js
         js_receipt,
         "cert",
         JS_NewString(ctx, receipt_out.cert.value().c_str()));
-    JS_SetPropertyStr(
-      ctx, js_receipt, "leaf", JS_NewString(ctx, receipt_out.leaf.c_str()));
+    if (receipt_out.leaf.has_value())
+    {
+      JS_SetPropertyStr(
+        ctx, js_receipt, "leaf", JS_NewString(ctx, receipt_out.leaf->c_str()));
+    }
+    if (receipt_out.leaf_components.has_value())
+    {
+      auto leaf_components = JS_NewObject(ctx);
+      if (receipt_out.leaf_components->write_set_digest.has_value())
```

Comment:
When can this be missing?

Reply:
Right now never, but there's benefit in keeping a single schema/set of json wrappers for receipts, and some specific app handlers could conceivably want to provide the full write set as a claim, so it seems sensible to have all the leaves optional.

Reply:
I don't understand, these are JS bindings, so the app is the js_generic app and it will always have those fields.

Reply:
After discussion, the agreement is to check for the following invariants before the objects are handed over to JS, so that TypeScript can model them accurately:

1. Either leaf or leaf_components must be present
2. All leaf_components must be present

There is no prejudice to a user wanting to mask or delete some components before serialisation.